### DISCUSSION TITLE
Combining Mesh & Plot in same axes

### DISCUSSION THREAD

**Discussion by akrisha20** [marcomusy/vedo#624] [Q&A] [answered]
Hi. I have imported an STL file as a Mesh, and I want to combine this Mesh with a 2D plot.
The 2D plot should be in the "YZ" direction.

I have created my plot by
`fig = plot(x,y,aspect=10).rotateX(90).rotateY(90)`

But when I add this to my overall plot using
`Plotter(shape=(1,1)).show(mesh,fig,at=0)`

The plot has a different axes from the mesh.
Since the data of the plot corresponds to a specific point in the mesh, I want to align these two.
I tried playing with fig.pos() & fig.scale to make the figure align with the mesh, but this was unsuccesfull.
Is there a way that I can add the plot to the mesh axes?

Many thanks!

**Comment by marcomusy** [comment]
Hi, if you use `aspect="equal"` axes should match. Example:
```python
from vedo import *
from vedo.pyplot import plot

settings.useParallelProjection = True # avoid perspective parallax

msh = Mesh(dataurl+"cessna.vtk")

x = np.linspace(-2,1, 20)
y = np.sin(x)
fig = plot(x, y, aspect="equal").shift(0,3.5)

plt = Plotter()
plt += [msh, Axes(msh), fig]
plt.show().close()
```
![Screenshot from 2022-04-11 21-58-08](https://user-images.githubusercontent.com/32848391/162819603-dde9af7b-624a-4ac3-923d-8e4b8b51f166.png)
(i just shifted the plot in y by 3.5 units to make it visible)

-----
PS:
to put it anywhere in the YZ plane:
```
fig = plot(x, y, aspect="equal").rotateX(90).rotateZ(90).shift(-2.5,0,2)
```
![Screenshot from 2022-04-11 22-08-40](https://user-images.githubusercontent.com/32848391/162823412-7a2b0aca-e8f4-4af5-b619-5f7a3ac2a4cc.png)

**Reply by akrisha20** [reply]
Thank you for the fast answer!
This seems promising, but by putting axis='equal', both the X and the Y axis are transformed.
I only want the "X axis" to be equal aspect with my mesh, because my Y values are very large
(In your example, imagine y = np.sin(x)*100)

**Comment by marcomusy** [comment, marked as answer]
x- axis is by default already at the same scaling of the world coordinates (where the 3D mesh lives).
In that case you don't need aspect="equal", just set the aspect ratio that you like (default is 4/3).
Not sure if that answers your question though..

**Reply by akrisha20** [reply]
Thank you. I was not aware of the same scaling. 
It works for now, by adjusting the aspect ratio.

Two other questions that popped up:
* I saw that there is an option for a logarithmic scale for the histograms. But I cannot find this option for the line plots. Does this exist?
* Also, is there an option for a single plot with two lines with different Y-axis? Such as the secondary axis in matplotlib

**Reply by marcomusy** [reply]
> I saw that there is an option for a logarithmic scale for the histograms. But I cannot find this option for the line plots. Does this exist?

Sorry it's not implemented in the scaling axis (vertical ticks and gridlines).. you can still plot the `np.log10(your_ydata)` instead of `your_data` and then declare it in the `ytitle` of the y axis (e.g. `plot(..., ytitle="log_10 (some intensity)`), so that it becomes linear. This is a perfectly valid practice.

**Reply by marcomusy** [reply]
> Also, is there an option for a single plot with two lines with different Y-axis? Such as the secondary axis in matplotlib

Yes! Just create an empty `Figure(xlim, ylim, axes=dict(xtitle=""))`
the empty xtitle will remove the x axis.

**Reply by akrisha20** [reply]
> > Also, is there an option for a single plot with two lines with different Y-axis? Such as the secondary axis in matplotlib
> 
> Yes! Just create an empty `Figure(xlim, ylim, axes=dict(xtitle=""))` the empty xtitle will remove the x axis.

But by doing that, your Y labels will overlap. Ideally, your y labels of the second plot should be on the other side to be visible

**Reply by marcomusy** [reply]
use `fig2.shift()`

**Comment by marcomusy** [comment]
I've added a new example in the gallery:
```python
"""Add a secondary y-axis for units conversion"""
from vedo import np, settings, dataurl, Mesh, show
from vedo.pyplot import plot, Figure

settings.annotatedCubeTexts = ['front','back','left','right','top','bttm']
msh = Mesh(dataurl+"cessna.vtk")

x0, x1 = [0.3, 2.0]
x = np.linspace(x0, x1, num=50)

# the main plot
fig1 = plot(
    x,
    1000*np.cos(x+1),
    xlim=[x0, x1],
    ylim=[-1000, 250],
    aspect=16/9,
    padding=0,
    title="Wing pull vs position",
    xtitle="Distance from central axis [m]",
    ytitle="N [Kg*m/s^2 ]",
)

# this empty Figure just creates a new y-axis
fig2 = Figure(
    fig1.xlim,                    # same as fig1
    np.array(fig1.ylim) * 7.236,  # conversion factor
    aspect=fig1.aspect,           # same as fig1
    padding=fig1.padding,         # same as fig1
    xtitle='',                    # don't draw x-axis!
    ytitle='Poundal [lb*ft/s^2 ]',
    axes=dict(                    # extra options for y-axis
        yShiftAlongX=1,           # shift 100% to the right
        yLabelOffset=-2,
        yLabelJustify="center-left",
        yTitlePosition=0.5,
        yTitleJustify="top-center",
        yTitleOffset=-0.15,
        axesLineWidth=3,
        c='red3',
    ),
)

fig1.rotateX(90).rotateZ(90).shift(-0.5, 0, 1)
fig2.rotateX(90).rotateZ(90).shift(-0.5, 0, 1)

cam = dict(
    pos=(4.119, -1.698, 1.423),
    focalPoint=(-0.2014, 0.6180, 0.2593),
    viewup=(-0.1925, 0.1297, 0.9727),
)
show(msh, fig1, fig2, __doc__, axes=5, camera=cam, bg2='lb').close()
```
![Screenshot from 2022-04-12 17-52-46](https://user-images.githubusercontent.com/32848391/163003530-19664019-ff0b-452e-b645-13981f6eea80.png)

**Reply by akrisha20** [reply]
Looks good!

If you would want to add data that corresponds with this second Y-axis, I imagine you have to use fig2 = plot instead of fig2 = Figure?
And use appropriate limits accordingly?

**Reply by marcomusy** [reply]
Correct!
